doc/implementation_details.rst

```diff
@@ -8,18 +8,117 @@ State machines
 * APMS        Acyclic Protocol Machine Sender
 * ALPMI       Alarm Protocol Machine Initiator
 * ALPMR       Alarm Protocol Machine Responder
+* CMDEV       Context Management Protocol Machine Device, handles connection establishment for IO-Devices.
+* CMRDR       Context Management Read Record Responder protocol machine device, responds to parameter read from the IO-controller
+* CMRPC       Context Management RPC Device Protocol Machine, handles RPC communication
+* CMSM        Context Management Surveillance Protocol Machine Device, monitors the establishment of a connection.
+* CPM         Consumer Protocol Machine, for receiving cyclic data
+* PPM         Cyclic Provider Protocol Machine
 
-Terminology:
 
-* APMX        General term for APMR and APMS
+CMRPC
+-----
+Handles the UDP communication in the start up phase, especially these
+messages:
 
+* connect
+* release
+* DControl ("Parameter end" is sent to IO-Device)
+* CControl ("Application ready" is sent to IO-Controller)
+* parameter read (Uses CMRDR)
+* parameter write
 
+Incoming UDP packets are parsed by ``pf_cmrpc_dce_packet()``, which also
+prepares the return UDP packet. This is done by putting together incoming
+fragments and then call ``pf_cmrpc_rpc_request()``.
 
+On DCE RPC connect requests the function ``pf_cmrpc_rm_connect_ind()`` is
+called, and it will create a DCE RPC connect response. It will also trigger
+these user callbacks:
 
-Messages
----------
+ * ``pnet_exp_module_ind()``
+ * ``pnet_exp_submodule_ind()``
+ * ``pnet_connect_ind()``
+ * ``pnet_state_ind()`` with PNET_EVENT_STARTUP
 
-* req
-* ind
-* rsp
-* cnf
\ No newline at end of file
+For incoming (parameter) write request messages is ``pf_cmrpc_rm_write_ind()`` called,
```

Comment:
pf_cmrpc_rm_write_ind() is called for incoming ....

Reply:
Done